provider/vsphere/internal/vsphereclient/client.go

```diff
@@ -0,0 +1,327 @@
+) (*Client, error) {
+	client, err := govmomi.NewClient(ctx, u, true)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &Client{client, datacenter, logger}, nil
+}
+
+// Close logs out and closes the client connection.
+func (c *Client) Close(ctx context.Context) error {
+	return c.client.Logout(ctx)
+}
+
+func (c *Client) recurser() *list.Recurser {
+	return &list.Recurser{
+		Collector: property.DefaultCollector(c.client.Client),
+		All:       true,
+	}
+}
+
+func (c *Client) finder(ctx context.Context) (*find.Finder, *object.Datacenter, error) {
+	finder := find.NewFinder(c.client.Client, true)
+	datacenter, err := finder.Datacenter(ctx, c.datacenter)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	finder.SetDatacenter(datacenter)
+	return finder, datacenter, nil
+}
+
+// RemoveVirtualMachines removes VMs matching the given path from the
+// system. The path may include wildcards, to match multiple VMs.
+func (c *Client) RemoveVirtualMachines(ctx context.Context, path string) error {
+	finder, _, err := c.finder(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	vms, err := finder.VirtualMachineList(ctx, path)
+	if err != nil {
+		if _, ok := err.(*find.NotFoundError); ok {
+			c.logger.Debugf("no VMs matching path %q", path)
+			return nil
+		}
+		return errors.Annotatef(err, "listing VMs at %q", path)
+	}
+
+	// We run all tasks in parallel, and wait for them below.
+	var lastError error
+	tasks := make([]*object.Task, 0, len(vms)*2)
+	for _, vm := range vms {
+		c.logger.Debugf("powering off %q", vm.Name())
+		task, err := vm.PowerOff(ctx)
+		if err != nil {
+			lastError = errors.Annotatef(err, "powering off %q", vm.Name())
+			c.logger.Errorf(err.Error())
+			continue
+		}
+		tasks = append(tasks, task)
+
+		c.logger.Debugf("destroying %q", vm.Name())
+		task, err = vm.Destroy(ctx)
+		if err != nil {
+			lastError = errors.Annotatef(err, "destroying %q", vm.Name())
+			c.logger.Errorf(err.Error())
+			continue
+		}
+		tasks = append(tasks, task)
+	}
+
+	for _, task := range tasks {
+		_, err := task.WaitForResult(ctx, nil)
+		if err != nil {
+			lastError = err
+			c.logger.Errorf(err.Error())
+		}
+	}
+	return errors.Annotate(lastError, "failed to remove instances")
+}
+
+// VirtualMachines return list of all VMs in the system matching the given path.
+func (c *Client) VirtualMachines(ctx context.Context, path string) ([]*mo.VirtualMachine, error) {
+	finder, _, err := c.finder(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	items, err := finder.VirtualMachineList(ctx, path)
+	if err != nil {
+		if _, ok := err.(*find.NotFoundError); ok {
+			return nil, nil
+		}
+		return nil, errors.Annotate(err, "listing VMs")
+	}
+
+	vms := make([]*mo.VirtualMachine, 0, len(items))
+	for _, item := range items {
+		var vm mo.VirtualMachine
+		err := c.client.RetrieveOne(ctx, item.Reference(), nil, &vm)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		vms = append(vms, &vm)
```

Comment:
I've been caught out in the past where setting the size and then appending to collection will actually have some empty items...
Maybe just use vms[i] =... here, especially since you could have an index on the loop if you changed _ to i?..